server/matchmaker/map_pool.py

```diff
@@ -0,0 +1,50 @@
+import random
+from collections import Counter
+from typing import List
+
+from ..decorators import with_logger
+from ..types import Map
+
+
+@with_logger
+class MapPool(object):
+    def __init__(
+        self,
+        map_pool_id: int,
+        name: str,
+        maps: List[Map] = []
+    ):
+        self.id = map_pool_id
+        self.name = name
+        self.set_maps(maps)
+
+    def set_maps(self, maps: List[Map]) -> None:
+        self.maps = {map_.id: map_ for map_ in maps}
+
+    def choose_map(self, played_map_ids: List[int] = []) -> Map:
+        """
+        Select a random map who's id does not appear in `played_map_ids`. If
+        all map ids appear in the list, then pick one that appears the least
+        amount of times.
+        """
+        if not self.maps:
+            self._logger.critical(
+                "Trying to choose a map from an empty map pool: %s", self.name
+            )
+            raise RuntimeError(f"Map pool {self.name} not set!")
+
+        # Make sure the counter has every available map
+        counter = Counter(self.maps.keys())
+        counter.update(id_ for id_ in played_map_ids if id_ in self.maps)
+
+        least_common = counter.most_common()[::-1]
+        least_count = least_common[0][1]
+        # Find the index of one past the last map with lowest count
+        for i, (_, count) in enumerate(least_common):
+            if count != least_count:
+                break
+        else:
+            i += 1
```

Comment:
Oh, you can probably just replace it by `i=len(least_common)`?

Reply:
No, because `least_common` is all of the maps, but sorted in reverse order by number of occurrences. As the comment says, we need this code in order to trim off all the maps which have higher play counts.

Reply:
Maybe I'm mistaken about the flow of execution, but I thought that the problem was potentially hitting `i += 1` when `least_common` is empty, since then `i` is not defined.
And as you only hit the `else` clause if you don't break, i.e. if every `count` is equal to `least_count` you will only ever hit the `i += 1` line with either `i` undefined or `i == len(least_common) - 1` after iterating through all of `least_common`, so might as well replace `i += 1` with `i = len(least_common)`?

Anyways I don't think it matters much with the empty check before, it's just to please whatever linter noraxx was using.

Reply:
Oh I see what you mean